core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -120,7 +120,7 @@ public AllocationService(Settings settings, AllocationDeciders allocationDecider
                                                           RoutingExplanations explanations) {
         final RoutingTable newRoutingTable = new RoutingTable.Builder().updateNodes(newRoutingNodes).build();
         MetaData newMetaData = updateMetaDataWithRoutingTable(oldMetaData, oldRoutingTable, newRoutingTable);
-        assert newRoutingTable.validate(newMetaData).valid();
+        newRoutingTable.validate(newMetaData); // validate the routing table is coherent with the cluster state metadata
```

Comment:
can we still do this under and assertion? (and yes, make it return true if all is OK). The reason is that we want to communicate that this is a sanity check and not really something we expect can happen here and want to guard against here